### PR TITLE
Add workaround for Intel in Sentinel

### LINKED ISSUE #2775: Nightly build errors intel/17 and intel/18 with Pthread,Serial: Kokkos_ThreadsExec.cpp(553): error: const variable "sentinel" requires an initializer
Changes possibly from #2715 are breaking intel/17.1.132 and 18.0.0.20170811  builds in the nightlies on Bowman (KNL arch) and intel/18.1.163 on Blake (Skylake)

**Error message**
```
[  6%] Building CXX object core/src/CMakeFiles/kokkoscore.dir/Threads/Kokkos_ThreadsExec_base.cpp.o
cd /home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/20200219025302/core/src && /home/projects/x86-64-knl/intel/compilers/2018/compilers_and_libraries_2018.0.128/linux/bin/intel64/icpc   -I/home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/20200219025302/core/src -I/home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/core/src -I/home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/20200219025302  -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -O3 -DNDEBUG   -xMIC-AVX512 -pthread -std=c++11 -o CMakeFiles/kokkoscore.dir/Threads/Kokkos_ThreadsExec_base.cpp.o -c /home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/core/src/Threads/Kokkos_ThreadsExec_base.cpp
/home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/core/src/Threads/Kokkos_ThreadsExec.cpp(553): error: const variable "sentinel" requires an initializer -- class "Kokkos::Impl::<unnamed>::Sentinel" has no user-provided default constructor
    static const Sentinel sentinel;
                                  ^

compilation aborted for /home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/core/src/Threads/Kokkos_ThreadsExec.cpp (code 2)
make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/Threads/Kokkos_ThreadsExec.cpp.o] Error 2
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/home/jenkins/bowman/workspace/Kokkos_SLURM_inner_test@2/20200219025302'
make[1]: *** [core/src/CMakeFiles/kokkoscore.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #2775. Also, see https://godbolt.org/z/Cuc4zS.

### CODE REVIEW

File: core/src/Threads/Kokkos_ThreadsExec.cpp [hunk 1]
```diff
@@ -78,6 +78,10 @@ void (*volatile s_current_function)(ThreadsExec &, const void *);
 const void *volatile s_current_function_arg = nullptr;
 
 struct Sentinel {
+#ifdef KOKKOS_COMPILER_INTEL
+  Sentinel() {}
+#endif
```
Comment (dalg24) [inline core/src/Threads/Kokkos_ThreadsExec.cpp]: Why are you not defaulting it?

Comment (masterleinad) [reply core/src/Threads/Kokkos_ThreadsExec.cpp]: Because that would still trigger the warning.

Comment (dalg24) [reply core/src/Threads/Kokkos_ThreadsExec.cpp]: And why did you chose this over `static Sentinel const sentinel{};`

Comment (masterleinad) [reply core/src/Threads/Kokkos_ThreadsExec.cpp]: I have not considered that so far, but it also seems to work.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Use an empty initializer instead (no `#ifdef`)

**Comment by masterleinad** [comment]
I changed it accordingly.

**Review by dalg24 — APPROVED** [review]

**Comment by ndellingwood** [comment]
I'll test this now in the failing configuration

**Comment by ndellingwood** [comment]
Changes in this PR were necessary to couple with #2774 to address the failures there. I confirm this eliminated the `sentinel`-related compilation errors in intel/17.0.098. I'm seeing some link errors in that build but that requires investigation separate from this PR.

**Pull request merged**